sips/sip-013/sip-013-semi-fungible-token-standard.md

```diff
@@ -0,0 +1,248 @@
+| `sft_mint`           | `{type: "sft_mint", token-id: uint, amount: uint, recipient: principal}`                        | Emitted when new tokens are minted.  |
+| `sft_burn`           | `{type: "sft_burn", token-id: uint, amount: uint, sender: principal}`                           | Emitted when tokens are burned.      |
+
+
+## Use of native asset functions
+
+Contract implementers should always use the built-in native assets that are provided as Clarity primitives whenever possible. This allows clients to use Post Conditions (explained below) and takes advantage of other benefits like native events and asset balances. However, there are no language primitives specific to semi-fungible tokens. The reference implementation included in this SIP therefore leverages the primitives to the extent that Clarity allows for.
+
+The recommended native asset primitives to use:
+
+- `define-fungible-token`
+- `ft-burn?`
+- `ft-get-balance`
+- `ft-get-supply`
+- `ft-mint?`
+- `ft-transfer?`
+- And their NFT equivalents.
+
+## Implementing in wallets and other applications
+
+Applications that interact with semi-fungible token contracts should validate if those contracts implement the SFT trait. If they do, then the application can use the interface described in this SIP for making transfers and getting other token information.
+
+All of the functions in this trait return the `response` type, which is a requirement of trait definitions in Clarity. However, some of these functions should be "fail-proof", in the sense that they should never return an error. The "fail-proof" functions are those that have been recommended as read-only. If a contract that implements this trait returns an error for these functions, it may be an indication of a faulty contract, and consumers of those contracts should proceed with caution.
+
+## Use of post conditions
+
+The Stacks blockchain includes a feature known as Post Conditions. By defining post conditions, users can create transactions that include pre-defined guarantees about what might happen in a contract. These post conditions can also be used to provide guarantees for custom fungible and non-fungible tokens that were defined using built-in Clarity primitives.
+
+There are no Clarity primitive counterparts for semi-fungible tokens, but contract developers can leverage a combination of post conditions to achieve the same result.
+
+There are two factor that should be checked by post conditions:
+
+1. The amount of semi-fungible tokens being transferred.
+2. The token ID of the semi-fungible token being transferred.
+
+To that end, it is recommended that developers use both Clarity primitives in their design. Semi-fungible token contracts can achieve complete post condition coverage by using both `define-fungible-token` and `define-non-fungible-token`.
+
+For strategies on how to best guard a semi-fungible token contract with post conditions, see the reference implementation at the end of this document.
```

Comment:
Can you also include a tarball of the source code as a supplemental file, per SIP-000?

Reply:
Also, I think it's in-scope how to represent SFTs as a combination of fungible and non-fungible tokens.  Even though we can't enforce this via a trait, wallets can require that the AST of the contract or even certain function implementations have certain properties (for example, this SIP could require that the core SFT mint/transfer/burn functionality adhere to a code _template_ which the wallet could inspect).  In addition, wallet implementations will still need to know exactly how to generate post-conditions on SFTs, and this is the perfect place to specify this.

Reply:
Yep that sounds great. I was actually looking at providing a `clarinet run` script that checks if a contract adheres to such a template. It looks like `clarinet` no longer exposes the AST which makes it a bit more tricky.

Reply:
Wait really?  That's annoying.  The node will report the contract's interface [1] right now, but not the AST.  I suppose the wallet could do some simple template-matching simply by fetching the contract source code and matching function bodies to a regex, but would get brittle really fast (but maybe that's what we want -- if the code isn't *exactly* what a regex says, then it's assumed nefarious).

[1] https://hirosystems.github.io/stacks-blockchain-api/#operation/get_contract_interface

Reply:
Yes. Would be pretty cool if the wallet can do it as well. I opened an issue for Clarinet here https://github.com/hirosystems/clarinet/issues/268. The mere ability to check if `ft-transfer?` is called at some point would already be a large benefit.